### ISSUE TITLE
Combine and streamline instructor audit logs and result tracking pages

### ISSUE THREAD

**Issue by wkurniawan07** [TEAMMATES/teammates#11812]
Current: we have two different activity-tracking pages built at different times, accessible under `/web/instructor/logs` and `/web/instructor/track-result-view`.

Problems:
- Both features are highly similar in both purpose and functionality. There is little reason to separate them to two different pages.
- Both features are only relevant in the context of a course. There is no such case as "I want to see the activity of my students in all my courses"; not only is the use case not valid(ated), but also the backend does not support that.
- `/logs` and `/track-result-view` are both ambiguous as a web address, especially so for `/logs`.

Solution:
- Both pages should be combined and accessible via `/web/instructor/courses/student-activity-logs`.
- Streamline the query panel.
  - Remove the course dropdown as it is no longer necessary. Change it to dropdown to choose the log type, e.g. "session access", "response submission", "result view".
  - Session name should be present in logs page and student name should be present in result tracking page. They are very valid use cases and in fact helps to reduce query time.
  - Result view tracking not being time bound is prone to performance problems.

**Comment by daongochieu2810** [comment]
>   * Session name should be present in logs page and student name should be present in result tracking page. They are very valid use cases and in fact helps to reduce query time.
 
@wkurniawan07 Can I ask what are the logs page and result tracking page referring to? We are having 1 page only right?

**Comment by damithc** [comment]
> * Remove the course dropdown as it is no longer necessary. Change it to dropdown to choose the log type, e.g. "session access", "response submission", "result view".

Perhaps use checkboxes so that the user can choose multiple log types? Alternatively, add another option in the drop-down 'session access and submission' as it is useful to see these two types of activities in a single timeline.

**Comment by daongochieu2810** [comment]
@damithc I agree with using checkboxes

**Comment by wkurniawan07** [comment]
> Perhaps use checkboxes so that the user can choose multiple log types

I mentioned before: no, because "result view" cannot be mixed with anything else.

**Comment by damithc** [comment]
> I mentioned before: no, because "result view" cannot be mixed with anything else.

Right. I remembered we discussed this before but couldn't find the source. 
@daongochieu2810 In that case, add the extra option to the drop-down.

**Comment by daongochieu2810** [comment]
@wkurniawan07 if we remove the course selection dropdown, do we show results from every course?

**Comment by wkurniawan07** [comment]
> Both features are only relevant in the context of a course. There is no such case as "I want to see the activity of my students in all my courses"; not only is the use case not valid(ated), but also the backend does not support that.

**Comment by damithc** [comment]
> * Remove the course dropdown as it is no longer necessary. Change it to dropdown to choose the log type, e.g. "session access", "response submission", "result view".

@wkurniawan07  I think @daongochieu2810 is referring to this suggestion. Isn't the course dropdown still required for the instructor to narrow down the scope to a course before choosing session name and student?
Or are you thinking of giving 'view student activities' option for each course in the instructor home page so that the course is already known when the user reaches this page?

**Comment by daongochieu2810** [comment]
@damithc This is my proposed UI, anything I need to change? Thanks

https://user-images.githubusercontent.com/53283766/172043276-72cf2d7c-7245-465f-a1dd-ee850950e2ce.mp4

**Comment by damithc** [comment]
> @damithc This is my proposed UI, anything I need to change? Thanks

@daongochieu2810 some thoughts:
The UI doesn't show which course it is.
Why there are two results tables? Show only the tables relevant to the activity type chosen.
If a session has been chosen, the results table can be expanded by default?

**Comment by daongochieu2810** [comment]
@damithc 
1. This page is accessed by clicking a button on a course table from the homepage, as you can see the course id on the search bar, but yes I will add the course name on the header
2. There are 3 options for the Activity Type dropdown: session access, response submission and both. The video shows me choosing the both option, so thats why there are 2 tables
3. The original audit log page only expand the table if it has no entries. I can change this behaviour if you want

**Comment by damithc** [comment]
> * This page is accessed by clicking a button on a course table from the homepage, as you can see the course id on the search bar, but yes I will add the course name on the header

Yes, must show prominently

> * There are 3 options for the Activity Type dropdown: session access, response submission and both. The video shows me choosing the both option, so thats why there are 2 tables

There should be 4 options:
* session access
* session submission
* session access and submission
* view session results

The first three are variations of previous audit logs page, and should be shown in the same table (i.e., one table per session) as they happen in the same timeline. If the user has chosen a session, these should be only one result table showing both access and submit activities of that session.

If the 4th option was chosen, it is still one table (but different columns).

Your views welcome; I may have misunderstood something.

> * The original audit log page only expand the table if it has no entries. I can change this behaviour if you want

OK to keep it. 'Expand if there is only one table' (because the user will want to expand it anyway) is a further optimization that will save the user from an extra click.

**Comment by daongochieu2810** [comment]
@damithc  isnt option 4 handled by `/web/instructor/sessions/report`?

**Comment by wkurniawan07** [comment]
@daongochieu2810 if option 4 is not there, you are not actually combining audit logs and result tracking pages; you were merely rewriting the audit logs page.

**Comment by daongochieu2810** [comment]
@damithc @wkurniawan07 Is this ok now?

https://user-images.githubusercontent.com/53283766/172180981-e2f7331e-add4-4918-9c55-c1d5207ddd01.mp4

**Comment by damithc** [comment]
> @damithc @wkurniawan07 Is this ok now?

Looks OK to me.

**Comment by wkurniawan07** [comment]
@damithc isn't "viewed session result" supposed to be displayed differently, i.e. it should display inactivity instead of activity?
Although, if you ask me, that should be a standalone display criteria. Wanting to view the lack of "viewed submission page" activity is a very valid use case.

**Comment by daongochieu2810** [comment]
@wkurniawan07 for 'viewed session result', isnt not displaying = inactivity?

**Comment by damithc** [comment]
> @damithc isn't "viewed session result" supposed to be displayed differently, i.e. it should display inactivity instead of activity?

@daongochieu2810 did you replicate in here the results view that is already in `/web/instructor/sessions/report`? Because we would have already discussed and decided a good display format when we implemented that feature earlier.

Also, posting screenshots for different use cases may be better than a video.

**Comment by daongochieu2810** [comment]
@damithc do you mean the one in `/web/instructor/track-result-view`?

**Comment by damithc** [comment]
Here is an example:
![image](https://user-images.githubusercontent.com/1673303/172210718-d9e0fabf-3298-4a54-a422-bf7690025463.png)

**Comment by wkurniawan07** [comment]
I'm just going to reiterate my previous point:

> [choosing to view activity or inactivity] should be a standalone display criteria. Wanting to view the lack of "viewed submission page" activity is a very valid use case [on top of viewing the lack of "viewed session results"].

**Comment by damithc** [comment]
> [choosing to view activity or inactivity] should be a standalone display criteria. Wanting to view the lack of "viewed submission page" activity is a very valid use case [on top of viewing the lack of "viewed session results"].

As a user, I'm more inclined to use activity for submission page and inactivity for results page. I'm also interested in know who didn't submit but that feature is in the session results page already. I haven't had a reason to find who did not view the submission page. 
Adding support for the 'missing' options (i e., an option to choose activity/inactivity) will make the UI more 'complete', but if we do, it's nice if the UI auto-selects the more common option by default (or else we end up making the common cases  harder-to-use by adding support for uncommon cases).

**Comment by daongochieu2810** [comment]
@damithc @wkurniawan07 so we add 2 checkboxes for active/inactive, and the active checkbox is checked by default for the 1st 3 options, and the inactive one is checked for the `view results` option?

**Comment by damithc** [comment]
> @damithc @wkurniawan07 so we add 2 checkboxes for active/inactive, and the active checkbox is checked by default for the 1st 3 options, and the inactive one is checked for the `view results` option?

@daongochieu2810 Doesn't the results view tracking table already show both last viewed and not-viewed but not-viewed on top? I can't check myself because I don't have a session that was active within the last 30 days.

**Comment by daongochieu2810** [comment]
@damithc yes thats true, but what about the other 3 log types from the old audit log page?

**Comment by damithc** [comment]
> @damithc yes thats true, but what about the other 3 log types from the old audit log page?

Does it make sense to similarly show the inactive cases at the bottom of the table? That way, the user doesn't have to deal with an additional setting. Does the data required is already available, without having to do more database reads?

**Comment by daongochieu2810** [comment]
so you would also want to see if the student hasnt submitted the response AND hasnt accessed the session? I think this is a valid use case

**Comment by damithc** [comment]
> so you would also want to see if the student hasnt submitted the response AND hasnt accessed the session? I think this is a valid use case

yes, this is the use case @wkurniawan07 raised. What I'm asking is if we can show that data _at the same time_ as what we show now, without additional cost to us and without the user needing to choose between active and inactive. If the user wants to see inactive students, just need to scroll to the bottom of the table.

**Comment by daongochieu2810** [comment]
ok I understand. Another issue I noticed is that we are logging multiple submission and access events, so 1 student can have a lot of these events. Should we take the latest event for each type instead?

**Comment by damithc** [comment]
> Another issue I noticed is that we are logging multiple submission and access events, so 1 student can have a lot of these events. Should we take the latest event for each type instead?

No, we should show all of them so that an instructor can compare them against the times the student said s/he accessed the session.

**Comment by wkurniawan07** [comment]
> No, we should show all of them so that an instructor can compare them against the times the student said s/he accessed the session.

@damithc this is asking for trouble. All it takes is one student spam-refreshing the submission page for the logs record to blow out of proportion.

**Comment by damithc** [comment]
> @damithc this is asking for trouble. All it takes is one student spam-refreshing the submission page for the logs record to blow out of proportion.

That's true. But I don't think we can avoid it entirely; the instructor might want to verify claims such as "I tried several times yesterday and again today". Also, hiding an access attempt before the deadline in favor of showing an attempt done after the deadline can give the wrong impression. Perhaps show accesses that are close together as `(multiple access)`?

**Comment by wkurniawan07** [comment]
I'm not asking to mindlessly collapse every access by one person into one log record; that obviously doesn't work. But we need to guard against spammy actions, and this is an attack vector with near-zero difficulty.
This should be kept in mind when designing #11843 (in fact, already mentioned there).

**Comment by daongochieu2810** [comment]
@wkurniawan07 we can add a new field `lastLogTimestamp` to `StudentAttributes`, and check that the minimum time difference with the next event must be > a threshold eg 2s

**Comment by daongochieu2810** [comment]
@damithc I have added the checkbox and this is how it looks now

![Screenshot from 2022-06-17 15-33-35](https://user-images.githubusercontent.com/53283766/174271973-810fd6d8-79fe-40ea-bdf1-7ee26098fdc0.png)
![Screenshot from 2022-06-17 16-36-25](https://user-images.githubusercontent.com/53283766/174271980-83876f38-1a05-467b-beb2-ec6d2dadcb9a.png)
![Screenshot from 2022-06-17 15-35-08](https://user-images.githubusercontent.com/53283766/174271987-ac21617b-5dda-4544-aad9-f4d5c33248f6.png)
![Screenshot from 2022-06-17 15-33-45](https://user-images.githubusercontent.com/53283766/174271990-c91a0e61-d69c-4753-9069-5c7b2fbe3627.png)
![Screenshot from 2022-06-17 15-33-42](https://user-images.githubusercontent.com/53283766/174271995-d505c4f1-fc8f-4917-9f00-a07b8d06de9d.png)

**Comment by damithc** [comment]
> @damithc I have added the checkbox and this is how it looks now

My thinking on this is that we don't need a separate setting for this. Instead, we just show both active and inactive in the same table. In case of results page tracking, show inactive on top; and in case of submission page, show active on top. In the less-common case of an instructor needing to see inactive session access or active result access, all she has to do is to scroll to the bottom of the table. But I'm open to hear if there are any downsides of taking that approach e.g., higher cost or lower usability

**Comment by wkurniawan07** [comment]
> lower usability

Any unnecessary info, situational or not, being spoon-fed to user (requiring scrolls etc.) is an outright negative score to usability.

**Comment by damithc** [comment]
> > lower usability
> 
> Any unnecessary info, situational or not, being spoon-fed to user (requiring scrolls etc.) is an outright negative score to usability.

Agreed, having to scroll or showing info the user did not actually need lowers usability. I have a similar (in fact, stronger) view of unnecessary configuration options being presented to users. Extra info at the bottom of a table is easier to ignore than extra configuration options at the top that every user will be forced to understand and use. Hence, I'm leaning towards not having this option but open to be persuaded otherwise.

**Comment by wkurniawan07** [comment]
> Extra info at the bottom of a table is easier to ignore than extra configuration options at the top that every user will be forced to understand and use.

Three problems with this statement:
- If the necessary info for the user is at the bottom of the page (remember that this page can show multiple sessions at once), there is an additional scrolling requirement if everything is shown.
- Users are in no way "forced to understand and use". Just because it is there doesn't mean anyone needs to use it.
- I am not sure that this is an unwanted functionality. I am no instructor, but wanting to know both action and lack of action in different situations/contexts sound like a plausible use case for me. With all due respect, "In case of results page tracking, show inactive on top; and in case of submission page, show active on top" sounds like a biased use case assumption to me.

**Comment by damithc** [comment]
> * If the necessary info for the user is at the bottom of the page (remember that this page can show multiple sessions at once), there is an additional scrolling requirement if everything is shown.

I agree, the additional scrolling can become quite troublesome in some cases e.g., large class and many sessions. And this got me thinking: should this option be given with the results (rather than with the input form) together with other related options e.g., 
`Expand/collapse all`  `[x] Show actions` `[x] Show inactions` 

The user is still given the power to choose but it is now easier to understand, ignore, and experiment with, and it might save us some round-trip calls to the server as well. But possibly to more complicated to implement? If you guys think this is less user-friendly or harder to implement, I'm fine with the current approach too. 

For the current approach, another alternative to consider is to use a drop down: 
`Show:`
* `Actions and inactions`
* `Actions only`
* `Inactions only`

> * I am not sure that this is an unwanted functionality. I am no instructor, but wanting to know both action and lack of action in different situations/contexts sound like a plausible use case for me. With all due respect, "In case of results page tracking, show inactive on top; and in case of submission page, show active on top" sounds like a biased use case assumption to me.

Yes, this is based on my own usage, which could be biased. My contention is the other use case is plausible but likely rare, and hence the UI should favor the common case while leaving the rare case also possible (i.e., not treat both common and rare case equally). But again, which is common and which is rare is just based on my speculation -- I could be wrong.

> * Users are in no way "forced to understand and use". Just because it is there doesn't mean anyone needs to use it.

Anyone using this page has to look at this option and at least decide it should be set or left alone, yes? If the option is not there at all, no such 'thinking' is needed.

**Comment by wkurniawan07** [comment]
> Anyone using this page has to look at this option and at least decide it should be set or left alone, yes? If the option is not there at all, no such 'thinking' is needed.

Sorry to say this, but the amount of settings we support in feedback sessions and feedback questions do not agree with this concern.

**Comment by damithc** [comment]
> > Anyone using this page has to look at this option and at least decide it should be set or left alone, yes? If the option is not there at all, no such 'thinking' is needed.
> 
> Sorry to say this, but the amount of settings we support in feedback sessions and feedback questions do not agree with this concern.

Ha ha, yes, we _do_ have a lot of configuration option (not entirely avoidable as we aim for high flexibility). Nevertheless, if any configuration option _can_ be omitted, we should do so. This applies to existing UIs too. If any of them can be simplified or optimized further for more common use cases, we should go for it.

**Comment by damithc** [comment]
To recap, the options so far:

[A] The current implementation of radio buttons in the input form.

[B1] Show tick boxes '[x] Show actions` and `[x] Show inactions` in the input form. 

[B2] Show tick boxes '[x] Show actions` and `[x] Show inactions` together with results. Ticking them will filter the outputs on the client side (i.e., no further requests to the server)

[C1] Show `Show actions and inactions`, `Show actions only`, `Show inactions only` as a drop-down, in the input form.

[C2] Similar to C1 but show with the results, and act on client-side.  

[D] No choice given to user. Always show both actions and inactions. Put inactions either at the top or the bottom of the table, depending on which is expected to be more common use case.

Any other ideas?

While I have a slight preference to D, I agree with the downsides pointed out and will not push for it. My next preference is B1, B2, C1, C2. I'm OK with any of them.
Lowest preference to A because it excludes the possibility of seeing both actions and inactions at the same time.

**Comment by daongochieu2810** [comment]
Ok I will go with B1 then

**Comment by wkurniawan07** [comment]
@daongochieu2810 why not B2?

**Comment by daongochieu2810** [comment]
@wkurniawan07 oh Im actually filtering the logs on the client side, so yes B2 is the simpler option, so I will go with that instead

**Comment by daongochieu2810** [comment]
![Screenshot from 2022-06-19 20-41-38](https://user-images.githubusercontent.com/53283766/174484433-10a2ced7-9e04-4cbe-b609-bf9aa2578930.png)
